File: src/lib/stores/meetingCost.ts
import { writable, derived, get } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';

export interface MeetingCostConfig {
  defaultHourlyRate: number;
  currencySymbol: string;
  currencyCode: string;
}

export interface MeetingCostState {
  running: boolean;
  paused: boolean;
  startedAt: number;
  pausedAt: number;
  totalPausedMs: number;
  attendees: number;
  hourlyRate: number;
  elapsedMs: number;
  totalCost: number;
  meetingsToday: number;
  totalCostToday: number;
  totalTimeTodayMs: number;
}

const DEFAULT_STATE: MeetingCostState = {
  running: false,
  paused: false,
  startedAt: 0,
  pausedAt: 0,
  totalPausedMs: 0,
  attendees: 2,
  hourlyRate: 75,
  elapsedMs: 0,
  totalCost: 0,
  meetingsToday: 0,
  totalCostToday: 0,
  totalTimeTodayMs: 0
};

function createMeetingCostStore() {
  const state = writable<MeetingCostState>({ ...DEFAULT_STATE });
  const config = writable<MeetingCostConfig>({
    defaultHourlyRate: 75,
    currencySymbol: '$',
    currencyCode: 'USD'
  });

  let tickInterval: ReturnType<typeof setInterval> | null = null;

  function startTicking() {
    stopTicking();
    tickInterval = setInterval(() => {
      const s = get(state);
      if (s.running && !s.paused) {
        const now = Date.now();
        const elapsed = now - s.startedAt - s.totalPausedMs;
        const hours = elapsed / 3_600_000;
        const cost = hours * s.attendees * s.hourlyRate;
        state.update(st => ({ ...st, elapsedMs: elapsed, totalCost: cost }));
      }
    }, 1000);
  }

  function stopTicking() {
    if (tickInterval) {
      clearInterval(tickInterval);
      tickInterval = null;
    }
  }

  return {
    subscribe: state.subscribe,
    config: { subscribe: config.subscribe },

    async init() {
      try {
        const [s, c] = await Promise.all([
          invoke<MeetingCostState>('meeting_get_state'),
          invoke<MeetingCostConfig>('meeting_get_config')
        ]);
        state.set(s);
        config.set(c);
        if (s.running) startTicking();
      } catch (error) {
        console.error('Failed to initialize meeting cost:', error);
      }
    },

    async start(attendees: number, hourlyRate?: number) {
      try {
        const result = await invoke<MeetingCostState>('meeting_start', { attendees, hourlyRate });
        state.set(result);
        startTicking();
      } catch (error) {
        console.error('Failed to start meeting:', error);
      }
    },

    async stop() {
      try {
        const result = await invoke<MeetingCostState>('meeting_stop');
        state.set(result);
        stopTicking();
      } catch (error) {
        console.error('Failed to stop meeting:', error);
      }
    },

    async pause() {
      try {
        const result = await invoke<MeetingCostState>('meeting_pause');
        state.set(result);
        stopTicking();
      } catch (error) {
        console.error('Failed to pause meeting:', error);
      }
    },

    async resume() {
      try {
        const result = await invoke<MeetingCostState>('meeting_resume');
        state.set(result);
        startTicking();
      } catch (error) {
        console.error('Failed to resume meeting:', error);
      }
    },

    async updateAttendees(attendees: number) {
      try {
        const result = await invoke<MeetingCostState>('meeting_update_attendees', { attendees });
        state.set(result);
      } catch (error) {
        console.error('Failed to update attendees:', error);
      }
    },

    async updateConfig(newConfig: MeetingCostConfig) {
      try {
        const result = await invoke<MeetingCostConfig>('meeting_set_config', { newConfig });
        config.set(result);
      } catch (error) {
        console.error('Failed to update meeting config:', error);
      }
    },

    async resetDaily() {
      try {
        const result = await invoke<MeetingCostState>('meeting_reset_daily');
        state.set(result);
      } catch (error) {
        console.error('Failed to reset daily stats:', error);
      }
    },

    cleanup() {
      stopTicking();
    }
  };
}

export const meetingCost = createMeetingCostStore();

export const meetingRunning = derived(
  { subscribe: meetingCost.subscribe },
  $state => $state.running
);

export const meetingPaused = derived(
  { subscribe: meetingCost.subscribe },
  $state => $state.paused
);

export function formatCost(amount: number, symbol: string = '$'): string {
  return `${symbol}${amount.toFixed(2)}`;
}

export function formatDuration(ms: number): string {
  const totalSec = Math.floor(ms / 1000);
  const hours = Math.floor(totalSec / 3600);
  const min = Math.floor((totalSec % 3600) / 60);
  const sec = totalSec % 60;
  if (hours > 0) {
    return `${hours}:${String(min).padStart(2, '0')}:${String(sec).padStart(2, '0')}`;
  }
  return `${min}:${String(sec).padStart(2, '0')}`;
}

export function costPerSecond(attendees: number, hourlyRate: number): number {
  return (attendees * hourlyRate) / 3600;
}

File: src-tauri/src/meetingcost.rs
//! Meeting Cost Timer - tracks meeting duration and calculates real-time cost
//!
//! Shows the running cost of a meeting based on number of attendees and
//! average hourly rate. Helps teams stay mindful of meeting time investment.

use serde::{Deserialize, Serialize};
use std::sync::Mutex;
use tauri::{AppHandle, Emitter, State};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeetingCostConfig {
    pub default_hourly_rate: f64,
    pub currency_symbol: String,
    pub currency_code: String,
}

impl Default for MeetingCostConfig {
    fn default() -> Self {
        Self {
            default_hourly_rate: 75.0,
            currency_symbol: "$".into(),
            currency_code: "USD".into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeetingCostState {
    pub running: bool,
    pub paused: bool,
    pub started_at: u64,
    pub paused_at: u64,
    pub total_paused_ms: u64,
    pub attendees: u32,
    pub hourly_rate: f64,
    pub elapsed_ms: u64,
    pub total_cost: f64,
    pub meetings_today: u32,
    pub total_cost_today: f64,
    pub total_time_today_ms: u64,
}

impl Default for MeetingCostState {
    fn default() -> Self {
        Self {
            running: false,
            paused: false,
            started_at: 0,
            paused_at: 0,
            total_paused_ms: 0,
            attendees: 2,
            hourly_rate: 75.0,
            elapsed_ms: 0,
            total_cost: 0.0,
            meetings_today: 0,
            total_cost_today: 0.0,
            total_time_today_ms: 0,
        }
    }
}

pub struct MeetingCostManager {
    config: Mutex<MeetingCostConfig>,
    state: Mutex<MeetingCostState>,
}

impl Default for MeetingCostManager {
    fn default() -> Self {
        Self {
            config: Mutex::new(MeetingCostConfig::default()),
            state: Mutex::new(MeetingCostState::default()),
        }
    }
}

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn calculate_cost(elapsed_ms: u64, attendees: u32, hourly_rate: f64) -> f64 {
    let hours = elapsed_ms as f64 / 3_600_000.0;
    hours * attendees as f64 * hourly_rate
}

#[tauri::command]
pub fn meeting_start(
    app: AppHandle,
    manager: State<'_, MeetingCostManager>,
    attendees: u32,
    hourly_rate: Option<f64>,
) -> Result<MeetingCostState, String> {
    let config = manager.config.lock().map_err(|e| e.to_string())?;
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;

    let rate = hourly_rate.unwrap_or(config.default_hourly_rate);
    let now = now_ms();

    state.running = true;
    state.paused = false;
    state.started_at = now;
    state.paused_at = 0;
    state.total_paused_ms = 0;
    state.attendees = attendees.max(1);
    state.hourly_rate = rate;
    state.elapsed_ms = 0;
    state.total_cost = 0.0;

    let _ = app.emit("meeting-started", state.clone());
    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_stop(
    app: AppHandle,
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;

    if !state.running {
        return Err("No meeting in progress".into());
    }

    let now = now_ms();
    let elapsed = now - state.started_at - state.total_paused_ms;
    let cost = calculate_cost(elapsed, state.attendees, state.hourly_rate);

    state.running = false;
    state.paused = false;
    state.elapsed_ms = elapsed;
    state.total_cost = cost;
    state.meetings_today += 1;
    state.total_cost_today += cost;
    state.total_time_today_ms += elapsed;

    let _ = app.emit("meeting-stopped", state.clone());
    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_pause(
    app: AppHandle,
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;

    if !state.running || state.paused {
        return Err("Meeting not running or already paused".into());
    }

    state.paused = true;
    state.paused_at = now_ms();

    let _ = app.emit("meeting-paused", state.clone());
    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_resume(
    app: AppHandle,
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;

    if !state.running || !state.paused {
        return Err("Meeting not paused".into());
    }

    let now = now_ms();
    state.total_paused_ms += now - state.paused_at;
    state.paused = false;
    state.paused_at = 0;

    let _ = app.emit("meeting-resumed", state.clone());
    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_update_attendees(
    manager: State<'_, MeetingCostManager>,
    attendees: u32,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;
    state.attendees = attendees.max(1);
    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_get_state(
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;

    if state.running && !state.paused {
        let now = now_ms();
        let elapsed = now - state.started_at - state.total_paused_ms;
        state.elapsed_ms = elapsed;
        state.total_cost = calculate_cost(elapsed, state.attendees, state.hourly_rate);
    }

    Ok(state.clone())
}

#[tauri::command]
pub fn meeting_get_config(
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostConfig, String> {
    let config = manager.config.lock().map_err(|e| e.to_string())?;
    Ok(config.clone())
}

#[tauri::command]
pub fn meeting_set_config(
    app: AppHandle,
    manager: State<'_, MeetingCostManager>,
    new_config: MeetingCostConfig,
) -> Result<MeetingCostConfig, String> {
    let mut config = manager.config.lock().map_err(|e| e.to_string())?;
    *config = new_config.clone();
    let _ = app.emit("meeting-config-changed", new_config.clone());
    Ok(new_config)
}

#[tauri::command]
pub fn meeting_reset_daily(
    manager: State<'_, MeetingCostManager>,
) -> Result<MeetingCostState, String> {
    let mut state = manager.state.lock().map_err(|e| e.to_string())?;
    state.meetings_today = 0;
    state.total_cost_today = 0.0;
    state.total_time_today_ms = 0;
    Ok(state.clone())
}
